test/cli/other_test.py

```diff
@@ -1213,18 +1214,134 @@ def test_multiple_define_rules(tmpdir):
 void f() { }
 ''')
 
-    exitcode, stdout, stderr = cppcheck(['--template=simple', '--rule-file={}'.format(rule_file), test_file])
+    exitcode, stdout, stderr = cppcheck(['--template=simple', '--rule-file={}'.format(rule_file), '-DDEF_3', test_file])
     assert exitcode == 0, stderr
     lines = stdout.splitlines()
     assert lines == [
         'Checking {} ...'.format(test_file),
         'Processing rule: DEF_1',
-        'Processing rule: DEF_2'
+        'Processing rule: DEF_2',
+        'Checking {}: DEF_3=1...'.format(test_file)
     ]
     lines = stderr.splitlines()
     assert lines == [
         "{}:2:0: error: found 'DEF_1' [ruleId1]".format(test_file),
-        "{}:3:0: error: found 'DEF_2' [ruleId2]".format(test_file)
+        "{}:3:0: error: define2 [ruleId2]".format(test_file)
+    ]
+
+
+def test_rule_file_define(tmpdir):
+    rule_file = os.path.join(tmpdir, 'rule_file.xml')
+    with open(rule_file, 'wt') as f:
+        f.write("""
+<rule>
+    <tokenlist>define</tokenlist>
+    <pattern>DEF_.</pattern>
+</rule>
+""")
+
+    test_file = os.path.join(tmpdir, 'test.c')
+    with open(test_file, 'wt') as f:
+        f.write('''
+#define DEF_1
+#define DEF_2
+void f() { }
+''')
+
+    exitcode, stdout, stderr = cppcheck(['--template=simple', '--rule-file={}'.format(rule_file), '-DDEF_3', test_file])
+    assert exitcode == 0, stdout
+    lines = stdout.splitlines()
+    assert lines == [
+        'Checking {} ...'.format(test_file),
+        'Processing rule: DEF_.',
+        'Checking {}: DEF_3=1...'.format(test_file)
+    ]
+    lines = stderr.splitlines()
+    assert lines == [
+        "{}:2:0: style: found 'DEF_1' [rule]".format(test_file),
+        "{}:3:0: style: found 'DEF_2' [rule]".format(test_file)
+    ]
+
+
+def test_rule_file_normal(tmpdir):
+    rule_file = os.path.join(tmpdir, 'rule_file.xml')
+    with open(rule_file, 'wt') as f:
+        f.write("""
+<rule>
+    <pattern>f</pattern>
+</rule>
+""")
+
+    test_file = os.path.join(tmpdir, 'test.c')
+    with open(test_file, 'wt') as f:
+        f.write('''
+#define DEF_1
+#define DEF_2
+void f() { }
+''')
+
+    exitcode, stdout, stderr = cppcheck(['--template=simple', '--rule-file={}'.format(rule_file), test_file])
+    assert exitcode == 0, stdout
+    lines = stdout.splitlines()
+    assert lines == [
+        'Checking {} ...'.format(test_file),
+        'Processing rule: f',
+    ]
+    lines = stderr.splitlines()
+    assert lines == [
+        "{}:4:0: style: found 'f' [rule]".format(test_file)
+    ]
+
+
+# TODO: what is the difference to "normal"
```

Comment:
I have not looked much into this - if somebody has some insight I can adjust the test.

Reply:
I guess the "raw" token list is probably the output from the preprocessor. i.e. I guess there is "typedef" tokens in the "raw" tokenstream.
we used to have "normal" tokens after basic tokenizer simplifications and "simplified" tokens after more aggressive tokenizer simplifications.

Reply:
Thanks for the explanation. Yes, `normal` seems to be what used to be `simplified`. I actually figured it out by myself a few days ago but didn't get around to update the test yet.

Reply:
I updated the tests.